bundles/org.openhab.core.model.rule.runtime/src/org/openhab/core/model/rule/runtime/internal/DSLRuleProvider.java

```diff
@@ -0,0 +1,432 @@
+import org.openhab.core.model.rule.rules.ChangedEventTrigger;
+import org.openhab.core.model.rule.rules.CommandEventTrigger;
+import org.openhab.core.model.rule.rules.EventEmittedTrigger;
+import org.openhab.core.model.rule.rules.EventTrigger;
+import org.openhab.core.model.rule.rules.GroupMemberChangedEventTrigger;
+import org.openhab.core.model.rule.rules.GroupMemberCommandEventTrigger;
+import org.openhab.core.model.rule.rules.GroupMemberUpdateEventTrigger;
+import org.openhab.core.model.rule.rules.RuleModel;
+import org.openhab.core.model.rule.rules.SystemOnShutdownTrigger;
+import org.openhab.core.model.rule.rules.SystemOnStartupTrigger;
+import org.openhab.core.model.rule.rules.ThingStateChangedEventTrigger;
+import org.openhab.core.model.rule.rules.ThingStateUpdateEventTrigger;
+import org.openhab.core.model.rule.rules.TimerTrigger;
+import org.openhab.core.model.rule.rules.UpdateEventTrigger;
+import org.openhab.core.model.script.runtime.DSLScriptContextProvider;
+import org.openhab.core.service.ReadyMarker;
+import org.openhab.core.service.ReadyMarkerFilter;
+import org.openhab.core.service.ReadyService;
+import org.openhab.core.service.ReadyService.ReadyTracker;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Reference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This RuleProvider provides rules that are defined in DSL rule files.
+ * All rules consist out of a list of triggers and a single script action.
+ * No rule conditions are used as this concept does not exist for DSL rules.
+ *
+ * @author Kai Kreuzer - Initial contribution
+ */
+@NonNullByDefault
+@Component(immediate = true, service = { DSLRuleProvider.class, RuleProvider.class, DSLScriptContextProvider.class })
+public class DSLRuleProvider
+        implements RuleProvider, ModelRepositoryChangeListener, DSLScriptContextProvider, ReadyTracker {
+
+    private static final String RULES_MODEL_NAME = "rules";
+    private static final String ITEMS_MODEL_NAME = "items";
+    private static final String THINGS_MODEL_NAME = "things";
+    static final String MIMETYPE_OPENHAB_DSL_RULE = "application/vnd.openhab.dsl.rule";
+
+    private final Logger logger = LoggerFactory.getLogger(DSLRuleProvider.class);
+    private final Collection<ProviderChangeListener<Rule>> listeners = new ArrayList<>();
+    private final Map<String, Rule> rules = new HashMap<>();
+    private final Map<String, IEvaluationContext> contexts = new HashMap<>();
+    private final Map<String, XExpression> xExpressions = new HashMap<>();
+    private int triggerId = 0;
+    private Set<String> markers = new HashSet<>();
+
+    @Reference
+    private @NonNullByDefault({}) ModelRepository modelRepository;
+
+    @Reference
+    private @NonNullByDefault({}) ReadyService readyService;
```

Comment:
Can you inject these fields using constructor injection instead? That allows for making them final and also makes it easier to write unit tests for this class without having to resort to reflection tricks.